### PR TITLE
DAOS-2548 Detect Fabric Interface Affinity

### PR DESCRIPTION
**Pull request by jbrosenz**
Detects all network devices and uses hwloc API
	to determine the affinity of each device found to
	facilitate binding in an upcoming update.

Signed-off-by: Joel Rosenzweig <joel.b.rosenzweig@intel.com>

### CODE REVIEW

**Comment by jbrosenz** [comment]
Working on unit tests and possible restructuring to accommodate the same.  Submitting the work in progress to get early eyeballs.

File: src/control/server/netdetect.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	var nodeset *C.char
+
+	err := C.hwloc_topology_init(&topology)
+	if err != C.int(0) {
+		return affinity, errors.New("hwloc_topology_init failure")
+	}
+
+	err = C.hwloc_topology_set_flags(topology,
+		C.HWLOC_TOPOLOGY_FLAG_IO_DEVICES)
+	if err != C.int(0) {
```
Comment (tanabarr) [inline src/control/server/netdetect.go]: could specify C.int(0) as some sort of constant if it's not already defined in the lib

File: src/control/server/netdetect.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		return affinity, errors.New("hwloc_topology_set_flags failure")
```
Comment (tanabarr) [inline src/control/server/netdetect.go]: we want to find a way of displaying the returned err in a meaningful way

File: src/control/server/netdetect.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	// store the cpuset and nodeset of the parent NUMA node or oldest
+	// non-root ancestor containing this object
+	for i = 0; i < numObj; i++ {
+		node = C.hwloc_get_obj_by_depth(topology, C.uint(depth), i)
+		if node != nil && strings.Contains(netNames, C.GoString(node.name)) {
+			ancestorNode := C.hwloc_get_ancestor_obj_by_type(topology,
+				C.HWLOC_OBJ_NUMANODE, node)
+			// If there is no NUMANODE ancestor, then look for the oldest
+			// non-root level ancestor (i.e. at depth 1)
+			if ancestorNode == nil {
```
Comment (tanabarr) [inline src/control/server/netdetect.go]: in what case will there not be a NUMANODE ancestor?

Comment (jbrosenz) [reply src/control/server/netdetect.go]: There are some system topologies that do not have NUMA nodes.  However, I found a single API that can replace this logic.  Using hwloc_get_non_io_ancestor_obj() will provide an ancestor node with a cpuset that is closest to the IO device node, and is suitable for binding.

**Comment by jbrosenz** [comment]
In this implementation, the high level function netdetectGetAffinityForNetworkDevices() is the high level entrypoint that the server can call to handle all of the work.  I'm working on the test functions now.  I'll update this PR with the test functions if we get agreement that this version of the implementation is what we want to go with.

File: src/control/server/netdetect.c [hunk 4]
```diff
@@ -0,0 +1,197 @@
+//
```
[the remark this answers is not in the document]
Comment (jbrosenz) [reply src/control/server/netdetect.c]: Wow, my editor setting on that file was spectacularly wrong!!  Please look at the content anyway while I fix the style issues.

**Comment by jbrosenz** [comment]
Addressed the checkpatch code style/warnings.  Unit tests are forthcoming, provided the content of the PR itself looks good to reviewers at this time.

**Pull request closed without merging**